Cleanup QgsDistanceArea API and sipify, add cache for ellipsoid lookup

nyalldawson opened the pull request:
Some 3.0 cleanups for the QgsDistanceArea api...

nyalldawson committed:
Cleanup QgsDistanceArea API

- remove methods which accept a QgsGeometry pointer, leave
duplicate methods which accept QgsGeometry reference (caller
should handle null geometry pointers in an acceptable
way instead of just returning 0 measurements)
- make protected members private. This class isn't designed
to be subclassed
- remove some methods not exposed to bindings and not used
anywhere

nyalldawson committed:
sipify QgsDistanceArea

nyalldawson committed:
Cleanup ellipsoidal mode detection in QgsDistanceArea

nyalldawson committed:
Use default copy constructors and assignment operators for QgsDistanceArea

There's no complex member variables here - the default implementations
should be fine (actually faster, since they won't require
recalculation of the derived parameters)

nyalldawson committed:
C++11 = good

nyalldawson committed:
Use a transparent cache to avoid db lookups when setting QgsDistanceArea ellipsoid

nyalldawson commented:
@wonder-sk I've added a cache for the ellipsoid lookup - can I request a re-review please? :)

wonder-sk reviewed (commented):
Caching of ellipsoids - oh yeah! Looks good to me, with one small note below...

I am wondering if this would be a good time to also throw away ```setEllipsoidalMode()``` method :-) Seems that calling ```setEllipsoid()``` is already good enough to distinguish whether we want planar or geodetic distances/areas

nyalldawson committed:
Fix possible corruption of cache

nyalldawson committed:
More QgsDistanceArea API cleanups and simplification

- setEllipsoidalMode() was removed. Ellipsoidal calculations are
now enabled whenever a valid ellipsoid() has been set. Set the
ellipsoid to 'NONE' to disable ellipsoidal calculations.

- ellipsoidalEnabled() was removed. Ellipsoidal calculations
are now enabled whenever a valid ellipsoid() is set. Check
willUseEllipsoid() to determine whether ellipsoidal
calculations will be performed.

- sourceCrs() and setSourceCrs() now always take and return
QgsCoordinateReferenceSystem objects. All string/ID based CRS
methods were removed.

- geographic() was removed. This was only used internally
in one place, so was replaced with sourceCrs().isGeographic() instead.

- some largely overlapping measurement functions were consolidated

nyalldawson commented on src/core/qgsdistancearea.cpp:
```diff
@@ -118,22 +85,35 @@ void QgsDistanceArea::setSourceAuthId( const QString &authId )
+  if ( cacheIt != sEllipsoidCache.constEnd() )
+  {
+    // found a match in the cache
+    sEllipsoidCacheLock.unlock();
```
Thanks - fixed!

nyalldawson commented:
> I am wondering if this would be a good time to also throw away setEllipsoidalMode() method :-) Seems that calling setEllipsoid() is already good enough to distinguish whether we want planar or geodetic distances/areas

Done, with a lot of other API cleanups too.

nyalldawson committed:
Move ellipsoid parameter retrieval to new QgsEllipsoidUtils class

nyalldawson committed:
Move retrieval of lists of known ellipsoids to QgsEllipsoidUtils

nyalldawson committed:
Add doxygen note linking QgsDistanceArea to QgsEllipsoidUtils
